### PR TITLE
User Management: Invite, Deactivate, Search, & Paginate

### PR DESCRIPTION
**Pull request by timeemit**
# Overview

This PR enables admin users to

- "Invite" users by allowing only whitelisted email addresses to sign up.
- "Deactivate" users by disallowing them from signing in or resetting their passwords.
- "Search" users that have either been invited or have signed up.
- "Paginate" large lists of users for organizations with > 10 users.

## Preview

Screenshot 2024-06-12 at 3 59 00 PM

Note: "Block" has been relabeled to "Deactivate" since opening the PR

## Technical Commentary

The invite list replaces the whitelist text file placed at */home/danswer_whitelist.txt* with a denormalized record created by the `dynamic_config_store` with a key of `"USERS"`.

The block list leverages [Fast API Users' `is_active` field](https://fastapi-users.github.io/fastapi-users/latest/configuration/schemas/?h=is_active).

## Tour of Functionality

https://github.com/danswer-ai/danswer/assets/874691/d9839912-ef84-436d-b72d-68474ed13256

Note: "Block" has been relabeled to "Deactivate" since opening the PR

### CODE REVIEW

File: backend/danswer/auth/invited_users.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from typing import cast
+
+from danswer.dynamic_configs.factory import get_dynamic_config_store
+from danswer.dynamic_configs.interface import ConfigNotFoundError
+from danswer.dynamic_configs.interface import JSON_ro
+
+USER_STORE_KEY = "USERS"
```
Comment (Weves) [inline backend/danswer/auth/invited_users.py]: nit: might be more descriptive to call this something like `INVITED_USERS` just so if someone is looking through the DB they can easily tell what this is

File: backend/danswer/server/manage/administrative.py [hunk 2]
```diff
@@ -231,3 +235,66 @@ def update_token_budget_settings(
...
+def remove_invited_user(
+    user_email: UserByEmail,
+    _: User | None = Depends(current_admin_user),
+) -> int:
+    user_emails = get_invited_users()
+    remaining_users = [user for user in user_emails if user != user_email.user_email]
+    return write_invited_users(remaining_users)
+
+
+@router.patch("/admin/block-user")
```
Comment (Weves) [inline backend/danswer/server/manage/administrative.py]: nit: prefer deactivate vs block

File: backend/danswer/server/manage/users.py [hunk 3]
```diff
@@ -69,11 +77,41 @@ async def demote_admin(
...
+                status=UserStatus.LIVE if user.is_active else UserStatus.BLOCKED,
+            )
+            for user in users
+        ][accepted_page * USERS_PAGE_SIZE : (accepted_page + 1) * USERS_PAGE_SIZE],
+        invited=[
+            InvitedUserSnapshot(email=email)
+            for email in invited_emails
+            if email not in accepted_emails
+        ][invited_page * USERS_PAGE_SIZE : (invited_page + 1) * USERS_PAGE_SIZE],
+        accepted_pages=accepted_count // USERS_PAGE_SIZE,
```
Comment (Weves) [inline backend/danswer/server/manage/users.py]: nit: should this be `accepted_count // USERS_PAGE_SIZE + 1`?

File: web/src/app/admin/users/page.tsx [hunk 4]
```diff
@@ -8,30 +15,46 @@ import {
...
 import useSWR, { mutate } from "swr";
+import useSWRMutation from "swr/mutation";
 import { ErrorCallout } from "@/components/ErrorCallout";
+import { HidableSection } from "@/app/admin/assistants/HidableSection";
+import BulkAdd from "@/components/admin/users/BulkAdd";
 
-const UsersTable = () => {
-  const { popup, setPopup } = usePopup();
+interface UsersResponse {
+  accepted: Array<User>;
```
Comment (Weves) [inline web/src/app/admin/users/page.tsx]: out of curiosity, any reason to use `Array<User>` vs `User[]`

File: web/src/components/admin/users/SignedUpUserTable.tsx [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+  block,
+  onSuccess,
+  onError,
+}: {
+  user: User;
+  block: boolean;
+  onSuccess: () => void;
+  onError: (message: string) => void;
+}) => {
+  const { trigger, isMutating } = useSWRMutation(
```
Comment (Weves) [inline web/src/components/admin/users/SignedUpUserTable.tsx:65]: haven't heard of this one, but it looks interesting 🧐

**Review by Weves — COMMENTED** [review]
Overall looks great!

File: web/src/app/admin/users/page.tsx (hunk 4 above)
Comment (thoughtcodex) [reply web/src/app/admin/users/page.tsx]: None, just a habit from previous projects.

File: backend/danswer/server/manage/users.py (hunk 3 above)
Comment (thoughtcodex) [reply backend/danswer/server/manage/users.py]: Yes!  Beautiful catch.

File: backend/danswer/auth/invited_users.py (hunk 1 above)
Comment (thoughtcodex) [reply backend/danswer/auth/invited_users.py]: Done: ed0c8533

File: backend/danswer/server/manage/administrative.py (hunk 2 above)
Comment (thoughtcodex) [reply backend/danswer/server/manage/administrative.py]: Done: f1b27239

File: backend/danswer/server/manage/users.py (hunk 3 above)
Comment (thoughtcodex) [reply backend/danswer/server/manage/users.py]: Done: 870fc7d6

File: web/src/app/admin/users/page.tsx (hunk 4 above)
Comment (thoughtcodex) [reply web/src/app/admin/users/page.tsx]: a072234a

**Pull request merged**